encrypt/aeadCipher.go

```diff
@@ -0,0 +1,376 @@
+package encrypt
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/md5"
+	"crypto/rand"
+	"crypto/sha1"
+	"io"
+
+	"golang.org/x/crypto/chacha20poly1305"
+	"golang.org/x/crypto/hkdf"
+)
+
+var (
+	// NonceSize defined the largest number nonce can reach, once reached, the nonce will warpped
+	NonceSize        = 12
+	readBufferSize   = 0xFFFFF             // get 4k for buffer size
+	hkdfInfoTag      = []byte("ss-subkey") // at present, this tag is defined as ss-subkey
+	nullNonce        = make([]byte, NonceSize, NonceSize)
+	aeadCipherMethod = map[string]*aeadGenerator{
+		"chacha20-ietf-poly1305": {32, 32, newChaCha20IETFEncoder},
+		"aes-256-gcm":            {32, 32, newAESGCMEncoder},
+		"aes-192-gcm":            {24, 24, newAESGCMEncoder},
+		"aes-128-gcm":            {16, 16, newAESGCMEncoder},
+	}
+)
+
+type aeadGenerator struct {
+	keyLen  int
+	saltLen int
+	newAEAD func(key, salt []byte, keylen int) (cipher.AEAD, error)
+}
+
+// increase uesd for nonce increacement, will warpped when reach the maximum of length byte can reach
+func increase(src []byte) {
+	for i := range src {
+		src[i]++
+		if src[i] != 0 {
+			return
+		}
+	}
+}
+
+// generate the salt randomly fror user
+func genSalt(salt []byte) {
+	// if salt is NULL, then gen the salt randomly
+	// by default, salt is random generated for each invocation by writer
+	// reader should input the specific salt to hkdf key
+	switch len(salt) {
+	case 16, 24, 32:
+	default:
+		panic("error salt len")
+	}
+	//FIXME
+	n, err := io.ReadFull(rand.Reader, salt)
+	if err != nil || n != len(salt) {
+		panic("error salt gen")
+	}
+}
+
+// key-derivation function for key pre-handle
+func kdf(password string, keyLen int) []byte {
+	var b, prev []byte
+	h := md5.New()
+	for len(b) < keyLen {
+		h.Write(prev)
+		h.Write([]byte(password))
+		b = h.Sum(b)
+		prev = b[len(b)-h.Size():]
+		h.Reset()
+	}
+	return b[:keyLen]
+}
+
+type aeadCipher struct {
+	psk        []byte
+	salt       []byte
+	readNonce  []byte
+	writeNonce []byte
+	readBuffer []byte
+	saltSize   int
+	keySize    int
+	dataLen    int
+
+	enc     cipher.AEAD
+	dec     cipher.AEAD
+	genator func(key, salt []byte, keylen int) (cipher.AEAD, error)
+}
+
+// NewAEADCipher creates a cipher that can be used in Dial() etc.
+// Use cipher.Copy() to create a new cipher with the same method and password
+// to avoid the cost of repeated cipher initialization.
+func NewAEADCipher(method, password string) (c Cipher, err error) {
+	if password == "" {
+		return nil, ErrEmptyPassword
+	}
+
+	var genator *aeadGenerator
+	genator, ok := aeadCipherMethod[method]
+	if !ok {
+		return nil, ErrUnsupportedMethod
+	}
+
+	cipher := aeadCipher{
+		readNonce:  make([]byte, NonceSize, NonceSize),
+		writeNonce: make([]byte, NonceSize, NonceSize),
+		salt:       make([]byte, genator.saltLen, genator.saltLen),
+		psk:        kdf(password, genator.keyLen), // per handle the passwd
```

Comment:
"per"  should be "pre"     
and does it necessary to be done before hkdf_sha1?

Reply:
https://github.com/shadowsocks/go-shadowsocks2/issues/60 
see this

Reply:
still, can't find the reason why perform kdf==
![image](https://user-images.githubusercontent.com/7588405/30107143-7c1f2d26-9330-11e7-81b2-bc11fac74149.png)
Is it this?

Reply:
maybe you can reopen this issue and ask riobard for help, He is the main designer of aead cipher. @dszhengyu